src/classes/BDI_AdditionalObjectService.cls

```diff
@@ -0,0 +1,391 @@
+/*
+    Copyright (c) 2019 Salesforce.org
+    All rights reserved.
+    
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+    
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+ 
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2019
+* @group Batch Data Import
+* @group-content ../../ApexDocContent/BatchDataImport.htm
+* @description Implements the ability to import additional objects during batch data import.
+*/
+
+public with sharing class BDI_AdditionalObjectService {
+
+    /*******************************************************************************************************
+    * @description the data import service we are invoked from
+    */
+    private BDI_DataImportService dataImportService;
+    private BDI_FieldMappingCustomMetadata fieldMapping;
```

Comment:
Would it be possible to set this up as a generic BDI_FieldMapping and define the necessary behaviors in the base class and implement them in both sub-classes? If the help text variant returns empty collections for those behaviors in a way that skips the additional object logic, that would allow us to execute this inline without having to test the type of the mapping class and branch accordingly

Reply:
The problem is that for the sake of efficiency I am attempting to query and sort out the data import custom metadata only once when the BDI_FieldMappingCustomMetadata is initially called, and not do it again in the additional object service, which means that I need those variables to be present and accessible in the version of the mapping class that gets to the AdditionalObjectService.

I could add those variables to base class, but that would be muddling the different implementations since it would be referring to metadata that the helptext mapping doesn't use.  In that situation I would then need to check whether those maps were populated before attempting to process.  In the end we are still doing a check, just on a different thing.